web/scripts/observations-backfill-calculated-cost.ts

```diff
@@ -0,0 +1,201 @@
+                ORDER BY models.project_id, models.start_date DESC NULLS LAST
+                LIMIT 1
+            ) m ON true
+            WHERE o.id > ${lastId} AND o.type = 'GENERATION' AND o.tmp_has_calculated_cost = FALSE
+            ORDER BY o.id
+            LIMIT ${batchSize}
+        ),
+        updated_batch AS (
+            UPDATE observations o
+            SET calculated_input_cost = COALESCE(batch.input_cost, batch.prompt_tokens::numeric * batch.input_price),
+                calculated_output_cost = COALESCE(batch.output_cost, batch.completion_tokens::numeric * batch.output_price),
+                calculated_total_cost = COALESCE(
+                    batch.total_cost,
+                    CASE
+                        WHEN batch.total_price IS NOT NULL AND batch.total_tokens IS NOT NULL THEN batch.total_price * batch.total_tokens::numeric
+                        ELSE batch.prompt_tokens::numeric * batch.input_price + batch.completion_tokens::numeric * batch.output_price
+                    END
+                ),
+                internal_model_id = batch.model_id,
+                tmp_has_calculated_cost = TRUE
+            FROM batch
+            WHERE o.id = batch.id
+            RETURNING o.id
+        )
+        -- Get the last id of the updated batch
+        SELECT id
+        FROM updated_batch
+        ORDER BY id DESC
+        LIMIT 1;
+      `);
+
+      log(`Batch update completed in ${Date.now() - startDate} ms`);
+
+      if (!batchUpdate[0]?.id) {
+        log(
+          `No more rows to process, breaking loop after ${totalRowsProcessed} rows processed.`,
+        );
+
+        break;
+      }
+
+      lastId = batchUpdate[0]?.id;
+      totalRowsProcessed += batchSize;
+
+      log(`Total rows processed after increment: ${totalRowsProcessed} rows`);
+      log(
+        `Progress: ${getPercentage(Math.min(totalRowsProcessed, Number(totalRowsToProcess)), Number(totalRowsToProcess))}`,
+      );
+
+      if (maxRowsToProcess && totalRowsProcessed >= maxRowsToProcess) {
+        log(`Max rows to process reached: ${maxRowsToProcess}, breaking loop.`);
+
+        break;
+      }
+
+      await new Promise((resolve) => setTimeout(resolve, sleepBetweenMs));
+    }
+
+    log("✅ Finished batch update loop.");
+
+    // Clean up
+    log("Dropping composite index...");
+    await prisma.$executeRaw`DROP INDEX IF EXISTS "observations_id_type_has_calculated_cost_idx";`;
+    log("✅ Dropped composite index");
+
+    // Drop the temporary column
+    log("Dropping temporary column...");
+    await prisma.$executeRaw`ALTER TABLE observations DROP COLUMN IF EXISTS tmp_has_calculated_cost;`;
+    log("✅ Dropped temporary column");
+
+    log("✅ Finished backfillCalculatedGenerationCost");
+  } catch (err) {
+    console.error("Error executing script", err);
+  } finally {
+    // Reset the statement timeout to two minutes
+    await prisma.$executeRaw`SET statement_timeout = '2min';`;
```

Comment:
For self hosters:
- lets store the variable somewhere and isnsert it back with raw unsave
- if it is 60 min it most likely means that we failed before and the script was triggered a second time. In this case, assume 2 min